qiita_pet/handlers/software.py

```diff
@@ -38,49 +63,83 @@ def get(self):
             active = False
 
         workflows = []
-        previous_outputs = []
         for w in DefaultWorkflow.iter(active=active):
             # getting the main default parameters
             nodes = []
             edges = []
-            for order, n in enumerate(w.graph.nodes):
-                dp = n.default_parameter
-                cmd = dp.command
-
-                # looping over the default parameters to make sure we got them
-                # all from required and optional parameters; whatever is left
-                # from required, are our inputs
-                rp = deepcopy(cmd.required_parameters)
-                op = deepcopy(cmd.optional_parameters)
-                params = dict()
-                for param, value in dp.values.items():
-                    if param in rp:
-                        del rp[param]
-                    if param in op:
-                        del op[param]
-                    params[param] = str(value)
-
-                # cmd_name, command id, command name,
-                # default params name, default parameters
-                cmd_name = 'command_%d' % order
-                nodes.append([cmd_name, cmd.id, cmd.name,
-                              dp.name, params])
-                for input in rp.values():
-                    accepted_values = ' | '.join(input[1])
-                    if order == 0:
-                        name = 'input_%d' % order
-                        nodes.append([name, cmd.id, accepted_values])
-                    else:
-                        name = 'output_%d_%s' % (order - 1, accepted_values)
-                    edges.append([name, cmd_name])
 
-                for output in cmd.outputs:
-                    previous_outputs.append(output[1])
-                    name = 'output_%d_%s' % (order, output[1])
-                    nodes.append([name, cmd.id, output[1]])
-                    edges.append([cmd_name, name])
+            # first get edges as this will give us the main connected commands
+            # and their order
+            graph = w.graph
+            # inputs is {input_type: node_name, ...} for easy look up of
+            # raw_inputs and reuse of the node_name
+            inputs = dict()
+            # main_nodes is {main_node_name: {
+            #                   output_type: output_node_name}, ...}
+            # for easy look up and merge of output_names
+            main_nodes = dict()
+            for x, y in graph.edges:
```

Comment:
I don't think I follow all that is going on in this loop. But my general comment would be to move this into a separate function that can be tested. Maybe as a class method or a function in the DefaultWorfklow module.